presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java

```diff
@@ -88,42 +92,26 @@ public Block readBlock(Type type)
                 // and use this as the skip size for the field readers
                 readOffset = presentStream.countBitsSet(readOffset);
             }
-            for (StreamReader structField : structFields) {
+            for (StreamReader structField : structFields.values()) {
                 structField.prepareNextRead(readOffset);
             }
         }
 
         List<Type> typeParameters = type.getTypeParameters();
 
         boolean[] nullVector = new boolean[nextBatchSize];
-        Block[] blocks = new Block[typeParameters.size()];
+        Block[] blocks;
+
         if (presentStream == null) {
-            for (int i = 0; i < typeParameters.size(); i++) {
-                if (i < structFields.length) {
-                    StreamReader structField = structFields[i];
-                    structField.prepareNextRead(nextBatchSize);
-                    blocks[i] = structField.readBlock(typeParameters.get(i));
-                }
-                else {
-                    blocks[i] = getNullBlock(typeParameters.get(i), nextBatchSize);
-                }
-            }
+            blocks = getBlocksForType(type, nextBatchSize);
         }
         else {
             int nullValues = presentStream.getUnsetBits(nextBatchSize, nullVector);
             if (nullValues != nextBatchSize) {
-                for (int i = 0; i < typeParameters.size(); i++) {
-                    if (i < structFields.length) {
-                        StreamReader structField = structFields[i];
-                        structField.prepareNextRead(nextBatchSize - nullValues);
-                        blocks[i] = structField.readBlock(typeParameters.get(i));
-                    }
-                    else {
-                        blocks[i] = getNullBlock(typeParameters.get(i), nextBatchSize - nullValues);
-                    }
-                }
+                blocks = getBlocksForType(type, nextBatchSize - nullValues);
             }
             else {
+                blocks = new Block[typeParameters.size()];
```

Comment:
Since `typeParameters` is only used in this branch move the declaration down to here.